File: ant-sdk/antd-php/examples/07-external-signer.php
<?php

/**
 * Example 07: External-signer flow — public file + single-chunk publish.
 *
 * PR #90 added prepareUploadPublic / finalizeUpload and prepareChunkUpload /
 * finalizeChunkUpload so the wallet key never has to live in the antd
 * daemon. This example uses anvil deterministic account #0 as the external
 * signer and exercises both round-trips end-to-end.
 *
 * See docs/external-signer-flow.md for the full reference; the IPaymentVault
 * function selector and tuple ABI are encoded inline via web3p/ethereum-abi.
 *
 * Requires:
 *   - web3p/ethereum-tx   (EIP-1559 tx signing)
 *   - web3p/ethereum-util (keccak256 + address derivation)
 *   - guzzlehttp/guzzle   (HTTP for JSON-RPC, already a runtime dep)
 */

require_once __DIR__ . '/../vendor/autoload.php';

use Autonomi\Antd\AntdClient;
use Web3p\EthereumTx\EIP1559Transaction;
use GuzzleHttp\Client as HttpClient;

// Anvil deterministic account #0. Pre-funded with ETH (gas) and antToken
// (storage payment) by `ant dev start --enable-evm` devnet genesis. Never
// use this key anywhere except a throw-away local devnet.
const ANVIL_KEY = 'ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80';
const ANVIL_ADDRESS = '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266';

// payForQuotes selector. See docs/external-signer-flow.md.
const PAY_FOR_QUOTES_SELECTOR = '0xb6c2141b';
// ERC-20 approve(address,uint256) selector.
const APPROVE_SELECTOR = '0x095ea7b3';

/**
 * Run approve + payForQuotes on-chain for a daemon prepare response.
 * Returns the quote_hash -> tx_hash map the daemon's finalize_* methods
 * expect. Every entry maps to the same payForQuotes tx because every
 * quote in the wave is paid in one batched call.
 *
 * @param  array<\Autonomi\Antd\Models\PaymentInfo>  $payments
 * @return array<string,string>
 */
function externalSignerPay(string $rpcUrl, string $vaultAddr, string $tokenAddr, array $payments): array
{
    // No on-chain work when every quoted chunk is already on-network.
    if (count($payments) === 0) {
        return [];
    }

    $http = new HttpClient(['base_uri' => $rpcUrl, 'timeout' => 30.0]);
    $sender = ANVIL_ADDRESS;
    $chainId = rpcCall($http, 'eth_chainId', [])['result'];
    $gasPrice = rpcCall($http, 'eth_gasPrice', [])['result'];

    // approve(vault, MAX) — idempotent and cheap; example uses MAX so
    // subsequent flows in this run skip a fresh approval.
    $approveData = APPROVE_SELECTOR
        . padAddress($vaultAddr)
        . padUint(str_repeat('f', 64));  // MAX_UINT256
    sendAndWait($http, $sender, $tokenAddr, $approveData, $chainId, $gasPrice, '0x' . dechex(500000));

    // payForQuotes((address rewardsAddress, uint256 amount, bytes32 quoteHash)[])
    // Encoding: [4-byte selector] [32-byte offset=0x20] [32-byte length=N]
    // [N x (address pad32 + amount pad32 + bytes32 pad32)]
    $count = count($payments);
    $body = padUint(dechex(0x20))   // offset to dynamic array
          . padUint(dechex($count)); // array length
    foreach ($payments as $p) {
        $amountHex = gmp_strval(gmp_init($p->amount, 10), 16);
        $qhHex = $p->quoteHash;
        if (str_starts_with($qhHex, '0x')) {
            $qhHex = substr($qhHex, 2);
        }
        $body .= padAddress($p->rewardsAddress)
              .  padUint($amountHex)
              .  $qhHex;  // bytes32 is already raw 32 bytes
    }
    $payData = PAY_FOR_QUOTES_SELECTOR . $body;
    $payTxHash = sendAndWait($http, $sender, $vaultAddr, $payData, $chainId, $gasPrice, '0x' . dechex(1000000));

    // Every quote in this wave was paid in the same call.
    $out = [];
    foreach ($payments as $p) {
        $out[$p->quoteHash] = $payTxHash;
    }
    return $out;
}

function rpcCall(HttpClient $http, string $method, array $params): array
{
    static $id = 0;
    $id++;
    $rsp = $http->post('', [
        'json' => ['jsonrpc' => '2.0', 'id' => $id, 'method' => $method, 'params' => $params],
        'headers' => ['Content-Type' => 'application/json'],
    ]);
    $j = json_decode($rsp->getBody()->getContents(), true);
    if (isset($j['error'])) {
        throw new \RuntimeException("$method error: " . json_encode($j['error']));
    }
    return $j;
}

function sendAndWait(HttpClient $http, string $from, string $to, string $data, string $chainIdHex, string $gasPriceHex, string $gasLimitHex): string
{
    $nonceHex = rpcCall($http, 'eth_getTransactionCount', [$from, 'pending'])['result'];

    // web3p/ethereum-tx's EIP1559Transaction RLP-encodes in canonical EIP-1559
    // field order. accessList must be present (empty list is fine) for the
    // RLP shape to be correct, otherwise anvil rejects with
    // "Failed to decode transaction".
    $tx = new EIP1559Transaction([
        'nonce'                => $nonceHex,
        'maxPriorityFeePerGas' => $gasPriceHex,
        'maxFeePerGas'         => '0x' . dechex(hexdec($gasPriceHex) * 2 + 1_000_000_000),
        'gasLimit'             => $gasLimitHex,
        'to'                   => $to,
        'value'                => '0x0',
        'data'                 => $data,
        'chainId'              => $chainIdHex,
        'accessList'           => [],
    ]);
    $signed = '0x' . $tx->sign(ANVIL_KEY);
    $txHash = rpcCall($http, 'eth_sendRawTransaction', [$signed])['result'];

    // Poll for receipt. Anvil instant-mines, so this typically resolves on
    // the first poll.
    for ($i = 0; $i < 600; $i++) {
        $r = rpcCall($http, 'eth_getTransactionReceipt', [$txHash]);
        if (isset($r['result']) && $r['result'] !== null) {
            if ($r['result']['status'] !== '0x1') {
                throw new \RuntimeException("tx $txHash reverted (status={$r['result']['status']})");
            }
            return $txHash;
        }
        usleep(100_000);  // 100 ms
    }
    throw new \RuntimeException("tx receipt timeout: $txHash");
}

function padAddress(string $addr): string
{
    $hex = strtolower(str_starts_with($addr, '0x') ? substr($addr, 2) : $addr);
    return str_repeat('0', 64 - strlen($hex)) . $hex;
}

function padUint(string $hex): string
{
    $hex = strtolower($hex);
    return str_repeat('0', 64 - strlen($hex)) . $hex;
}

// --- main ---

$tmp = sys_get_temp_dir() . DIRECTORY_SEPARATOR . uniqid('antd-php-07-extsig-');
mkdir($tmp, 0o700, true);
$client = new AntdClient();

try {
    // --- 1. file upload via external signer -------------------------------
    $srcFile = $tmp . '/file.bin';
    $fileContent = str_repeat("hello external signer from php (file)\n", 16);
    file_put_contents($srcFile, $fileContent);

    $filePrep = $client->prepareUploadPublic($srcFile);
    printf(
        "File prepare: upload_id=%s..., payment_type=%s, payments=%d, total_amount=%s\n",
        substr($filePrep->uploadId, 0, 16),
        $filePrep->paymentType,
        count($filePrep->payments),
        $filePrep->totalAmount
    );

    $fileTxHashes = externalSignerPay(
        $filePrep->rpcUrl,
        $filePrep->paymentVaultAddress,
        $filePrep->paymentTokenAddress,
        $filePrep->payments
    );
    $fileFin = $client->finalizeUpload($filePrep->uploadId, $fileTxHashes);
    printf(
        "File finalize: data_map_address=%s, chunks_stored=%d\n",
        $fileFin->dataMapAddress,
        $fileFin->chunksStored
    );

    $dstFile = $srcFile . '.downloaded';
    $client->fileGetPublic($fileFin->dataMapAddress, $dstFile);
    if (file_get_contents($dstFile) !== $fileContent) {
        fwrite(STDERR, "file round-trip mismatch\n");
        exit(1);
    }
    echo "File round-trip OK!\n";

    // --- 2. single-chunk publish via external signer ----------------------
    $chunkData = str_repeat("hello external signer from php (chunk)\n", 8);
    $chunkPrep = $client->prepareChunkUpload($chunkData);
    if ($chunkPrep->alreadyStored) {
        printf("Chunk prepare: already_stored, address=%s\n", $chunkPrep->address);
    } else {
        printf(
            "Chunk prepare: upload_id=%s..., address=%s, payments=%d, total_amount=%s\n",
            substr($chunkPrep->uploadId, 0, 16),
            $chunkPrep->address,
            count($chunkPrep->payments),
            $chunkPrep->totalAmount
        );
        $chunkTxHashes = externalSignerPay(
            $chunkPrep->rpcUrl,
            $chunkPrep->paymentVaultAddress,
            $chunkPrep->paymentTokenAddress,
            $chunkPrep->payments
        );
        $addr = $client->finalizeChunkUpload($chunkPrep->uploadId, $chunkTxHashes);
        if ($addr !== $chunkPrep->address) {
            fwrite(STDERR, "chunk address mismatch: {$addr} != {$chunkPrep->address}\n");
            exit(1);
        }
        printf("Chunk finalize: address=%s\n", $addr);
    }

    $chunkGot = $client->chunkGet($chunkPrep->address);
    if ($chunkGot !== $chunkData) {
        fwrite(STDERR, "chunk round-trip mismatch\n");
        exit(1);
    }
    echo "Chunk round-trip OK!\n";

    echo "\n07_external_signer OK!\n";
} finally {
    array_map('unlink', glob($tmp . '/*') ?: []);
    @rmdir($tmp);
}

File: ant-sdk/antd-rust/examples/07-external-signer.rs
//! Example 07: External-signer flow — public file + single-chunk publish.
//!
//! PR #90 added `prepare_upload_public` / `finalize_upload` and
//! `prepare_chunk_upload` / `finalize_chunk_upload` so the wallet key
//! never has to live in the antd daemon. This example uses anvil
//! deterministic account #0 as the external signer and exercises both
//! round-trips end-to-end.
//!
//! See `docs/external-signer-flow.md` for the full reference; the
//! `IPaymentVault` contract bindings are baked in via alloy's `sol!`
//! macro from the JSON ABI committed at `docs/abi/IPaymentVault.json`.

use std::collections::HashMap;
use std::fs;

use alloy::network::EthereumWallet;
use alloy::primitives::{Address, FixedBytes, U256};
use alloy::providers::ProviderBuilder;
use alloy::signers::local::PrivateKeySigner;
use alloy::sol;
use antd_client::{Client, DEFAULT_BASE_URL};

// Anvil deterministic account #0. Pre-funded with ETH (gas) and antToken
// (storage payment) by `ant dev start --enable-evm` devnet genesis. Never
// use this key anywhere except a throw-away local devnet.
const ANVIL_KEY: &str = "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";

sol! {
    #[allow(missing_docs)]
    #[sol(rpc)]
    contract IERC20 {
        function approve(address spender, uint256 value) external returns (bool);
    }

    #[allow(missing_docs)]
    #[sol(rpc)]
    contract IPaymentVault {
        struct DataPayment {
            address rewardsAddress;
            uint256 amount;
            bytes32 quoteHash;
        }
        function payForQuotes(DataPayment[] memory payments) external;
    }
}

/// Run approve + payForQuotes on-chain for a daemon prepare response.
/// Returns the `quote_hash -> tx_hash` map the daemon's `finalize_*`
/// methods expect. Every entry maps to the same `payForQuotes` tx
/// because every quote in the wave is paid in one batched call.
async fn external_signer_pay(
    rpc_url: &str,
    vault_addr: Address,
    token_addr: Address,
    payments: &[antd_client::PaymentInfo],
    signer: &PrivateKeySigner,
) -> Result<HashMap<String, String>, Box<dyn std::error::Error>> {
    // No on-chain work when every quoted chunk is already on-network.
    if payments.is_empty() {
        return Ok(HashMap::new());
    }

    let wallet = EthereumWallet::from(signer.clone());
    let provider = ProviderBuilder::new()
        .wallet(wallet)
        .connect_http(rpc_url.parse()?);

    // approve(vault, MAX) -- idempotent and cheap; example uses MAX so
    // subsequent flows in this run skip a fresh approval.
    let token = IERC20::new(token_addr, provider.clone());
    let approve_receipt = token
        .approve(vault_addr, U256::MAX)
        .send()
        .await?
        .get_receipt()
        .await?;
    if !approve_receipt.status() {
        return Err(format!("approve reverted: {approve_receipt:?}").into());
    }

    // payForQuotes -- one tx covering every quote in this wave.
    let vault = IPaymentVault::new(vault_addr, provider.clone());
    let data_payments: Vec<IPaymentVault::DataPayment> = payments
        .iter()
        .map(|p| -> Result<_, Box<dyn std::error::Error>> {
            let rewards_address: Address = p.rewards_address.parse()?;
            let amount: U256 = p.amount.parse()?;
            let qh_hex = p.quote_hash.strip_prefix("0x").unwrap_or(&p.quote_hash);
            let qh = FixedBytes::<32>::from_slice(&hex::decode(qh_hex)?);
            Ok(IPaymentVault::DataPayment {
                rewardsAddress: rewards_address,
                amount,
                quoteHash: qh,
            })
        })
        .collect::<Result<Vec<_>, _>>()?;

    let pay_receipt = vault
        .payForQuotes(data_payments)
        .send()
        .await?
        .get_receipt()
        .await?;
    if !pay_receipt.status() {
        return Err(format!("payForQuotes reverted: {pay_receipt:?}").into());
    }
    let pay_tx_hash = format!("{:#x}", pay_receipt.transaction_hash);

    // Every quote in this wave was paid in the same call.
    let mut tx_hashes = HashMap::new();
    for p in payments {
        tx_hashes.insert(p.quote_hash.clone(), pay_tx_hash.clone());
    }
    Ok(tx_hashes)
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let client = Client::new(DEFAULT_BASE_URL);
    let signer: PrivateKeySigner = ANVIL_KEY.parse()?;

    let tmp = std::env::temp_dir().join("antd-rust-07-external-signer");
    let _ = fs::remove_dir_all(&tmp);
    fs::create_dir_all(&tmp)?;

    // --- 1. file upload via external signer -------------------------------
    let src = tmp.join("file.bin");
    let file_content = "hello external signer from rust (file)\n".repeat(16); // ~624 bytes
    fs::write(&src, &file_content)?;

    let file_prep = client.prepare_upload_public(src.to_str().unwrap()).await?;
    println!(
        "File prepare: upload_id={}..., payment_type={}, payments={}, total_amount={}",
        &file_prep.upload_id[..16],
        file_prep.payment_type,
        file_prep.payments.len(),
        file_prep.total_amount,
    );

    let tx_hashes = external_signer_pay(
        &file_prep.rpc_url,
        file_prep.payment_vault_address.parse()?,
        file_prep.payment_token_address.parse()?,
        &file_prep.payments,
        &signer,
    )
    .await?;
    let fin = client
        .finalize_upload(&file_prep.upload_id, &tx_hashes)
        .await?;
    println!(
        "File finalize: data_map_address={}, chunks_stored={}",
        fin.data_map_address, fin.chunks_stored,
    );

    let dst = tmp.join("file.bin.downloaded");
    client
        .file_get_public(&fin.data_map_address, dst.to_str().unwrap())
        .await?;
    let got = fs::read(&dst)?;
    if got != file_content.as_bytes() {
        return Err("file round-trip mismatch".into());
    }
    println!("File round-trip OK!");

    // --- 2. single-chunk publish via external signer ----------------------
    let chunk_data = "hello external signer from rust (chunk)\n"
        .repeat(8)
        .into_bytes();
    let chunk_prep = client.prepare_chunk_upload(&chunk_data).await?;
    if chunk_prep.already_stored {
        println!(
            "Chunk prepare: already_stored, address={}",
            chunk_prep.address
        );
    } else {
        println!(
            "Chunk prepare: upload_id={}..., address={}, payments={}, total_amount={}",
            &chunk_prep.upload_id[..16],
            chunk_prep.address,
            chunk_prep.payments.len(),
            chunk_prep.total_amount,
        );
        let tx_hashes = external_signer_pay(
            &chunk_prep.rpc_url,
            chunk_prep.payment_vault_address.parse()?,
            chunk_prep.payment_token_address.parse()?,
            &chunk_prep.payments,
            &signer,
        )
        .await?;
        let addr = client
            .finalize_chunk_upload(&chunk_prep.upload_id, &tx_hashes)
            .await?;
        if addr != chunk_prep.address {
            return Err(
                format!("chunk address mismatch: {} != {}", addr, chunk_prep.address).into(),
            );
        }
        println!("Chunk finalize: address={addr}");
    }

    let got = client.chunk_get(&chunk_prep.address).await?;
    if got != chunk_data {
        return Err("chunk round-trip mismatch".into());
    }
    println!("Chunk round-trip OK!");

    fs::remove_dir_all(&tmp).ok();
    println!("\n07-external-signer OK!");
    Ok(())
}
